### ISSUE TITLE
Clarify and standardize `RESTRICT_ON_SEND` type

### ISSUE THREAD

**Issue by viralpraxis** [rubocop/rubocop#14163]
At the moment, the `RESTRICT_ON_SEND` constant is not standardized:

```bash
$ bundle exec ruby -rrubocop -e 'p ObjectSpace.each_object(Class).select { it.<(RuboCop::Cop::Base) }.map { it::RESTRICT_ON_SEND }.map(&:class).tally' 
{Set => 434, Array => 146}
```

While `RuboCop::Cop::Base::RESTRICT_ON_SEND` is an empty set by default, some cops override it with a non-empty set:

```bash
$ bundle exec ruby -rrubocop -e 'p ObjectSpace.each_object(Class).select { it.<(RuboCop::Cop::Base) }.select { it::RESTRICT_ON_SEND.is_a?(Set) && it::RESTRICT_ON_SEND.any? }'
[RuboCop::Cop::Style::YodaCondition, RuboCop::Cop::Style::FileWrite, RuboCop::Cop::Style::RedundantFormat, RuboCop::Cop::Lint::DuplicateRequire, RuboCop::Cop::Lint::RedundantTypeConversion, RuboCop::Cop::Lint::OutOfRangeRegexpRef]
```

I'd like to propose using Set consistently everywhere. This should be a backward-compatible change, and it better reflects the semantics of the constant. Additionally, it would allow static analysis tools (such as `Lint/DuplicatedSetElement`) to provide useful feedback.

I do not expect any performance impact since AFAIK internally small MRI sets (hashes) are built upon arrays

We can also introduce a new `InternalAffairs` cop to help plugins transition.

**Comment by Earlopain** [comment]
I could be wrong, but I'm pretty sure there's a `to_set` somewhere that converts everything to a set already. If so, practically it shouldn't make any difference

**Comment by viralpraxis** [comment]
> I could be wrong, but I'm pretty sure there's a `to_set` somewhere that converts everything to a set already. If so, practically it shouldn't make any difference

There is, in fact, a `to_a` conversion:

https://github.com/rubocop/rubocop/blob/416d344cedc60e9a839b2164edb319a3d2d6a973/lib/rubocop/cop/base.rb#L402

**Comment by Earlopain** [comment]
Oops, other way around then 😅

**Comment by viralpraxis** [comment]
Yeah, I understand that it doesn't make much difference in most use cases (is the worst-case scenario we will just perform some redundant computation). I'm mostly bringing this up because:

1. `RESTRICT_ON_SEND` is part of public developer API
2. It's not obvious (from the [docs](https://docs.rubocop.org/rubocop/development.html) and auto-generated templates) that RuboCop developers can use a `Set` instead of an `Array` if they want to.
3. It's also unclear (to me) whether it would be acceptable to use `RESTRICT_ON_SEND = (...).to_set` in a cop proposed for inclusion in RuboCop core or plugin.

To elaborate on p.2: I prefer to wrap collection literals in the most appropriate container type -- and in this case, `Set` seems like the best fit for that constant.

Perhaps it would be enough to add a note to the developer documentation?
